### PR TITLE
ChannelKeys - separate commitment revocation from getting the per-commitment point

### PR DESCRIPTION
**Pull request by devrandom**
The commitment secret is sensitive - it can be used by an attacker to steal funds if the node also signs the same transaction. In general, the signer should be considered more secure - it will, for example, be implemented in secure hardware. It should not give the rest of the node access to data which can be used by an attacker and it should enforce policy which prevents attacks.

### CODE REVIEW

File: lightning/src/chain/keysinterface.rs [hunk 1]
```diff
@@ -195,9 +195,12 @@ impl Readable for SpendableOutputDescriptor {
...
 // ChannelMonitors instead of expecting to clone the one out of the Channel into the monitors.
 pub trait ChannelKeys : Send+Clone {
-	/// Gets the commitment seed for a specific commitment number
+	/// Gets the per-commitment point for a specific commitment number
 	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards
-	fn commitment_secret(&self, idx: u64) -> [u8; 32];
+	fn get_per_commitment_point<T: secp256k1::Signing + secp256k1::Verification>(&self, idx: u64, secp_ctx: &Secp256k1<T>) -> PublicKey;
+	/// Gets the commitment seed for a specific commitment number, thereby revoking the commitment
+	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards
+	fn revoke_commitment(&self, idx: u64) -> [u8; 32];
```
Comment (valentinewallace) [inline lightning/src/chain/keysinterface.rs:203]: Hm, given this function is also used to get the secret for channel_reestablish, not just revocation, seems the original name `commitment_secret` (or maybe `get_commitment_secret` or so) might be more appropriate?

Comment (TheBlueMatt) [reply lightning/src/chain/keysinterface.rs:203]: Or maybe get_revoke_commitment_secret or so, to indicate that its both revoking and that it may be called twice for the same commitment in the reestablish case - either way, worth nothing that this may happen in the docs.

File: lightning/src/util/enforcing_trait_impls.rs [hunk 2]
```diff
@@ -48,7 +48,11 @@ impl EnforcingChannelKeys {
 }
 
 impl ChannelKeys for EnforcingChannelKeys {
-	fn commitment_secret(&self, idx: u64) -> [u8; 32] { self.inner.commitment_secret(idx) }
+	fn get_per_commitment_point<T: secp256k1::Signing + secp256k1::Verification>(&self, idx: u64, secp_ctx: &Secp256k1<T>) -> PublicKey {
+		self.inner.get_per_commitment_point(idx, secp_ctx)
+	}
+
+	fn revoke_commitment(&self, idx: u64) -> [u8; 32] { self.inner.revoke_commitment(idx) }
```
Comment (TheBlueMatt) [inline lightning/src/util/enforcing_trait_impls.rs:55]: Out of curiosity, do all the tests still pass if we require that we never re-sign a commitment transaction after revoking its index? Seems like thats the right next step to enforcement here.

File: lightning/src/chain/keysinterface.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Using Matt's name, which I think attracts attention to the fact that this is not a pure function, but also has some policy enforcement side effects.

File: lightning/src/util/enforcing_trait_impls.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/util/enforcing_trait_impls.rs]: I pushed a commit for that, but `test_fee_spike_violation_fails_htlc` is doing something out of order, even after an attempted fix.  I can try to correct this - may need to move things around in the test.  Let me know what you think of the direction.

**Comment by devrandom** [comment]
@valentinewallace the motivation is that the commitment secret is sensitive - it can be used by an attacker to steal funds if the node also signs the same transaction.  In general, the signer should be considered more secure - it will, for example, be implemented in secure hardware.  It should not give the rest of the node access to data which can be used by an attacker and it should enforce policy which prevents attacks.

In this particular case, some of the policies include: revoke commitments in order, don't sign a revoked transaction, don't revoke a signed transaction.  In the extra WIP commit I implemented the first two.

File: lightning/src/util/enforcing_trait_impls.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/util/enforcing_trait_impls.rs]: I force-pushed a fix for that, should pass now.

File: lightning/src/chain/keysinterface.rs [hunk 3]
```diff
@@ -217,6 +225,7 @@ pub trait ChannelKeys : Send+Clone {
 	/// Create a signature for a local commitment transaction. This will only ever be called with
 	/// the same local_commitment_tx (or a copy thereof), though there are currently no guarantees
 	/// that it will not be called multiple times.
+	/// The commitment must not have been revoked.
```
Comment (ariard) [inline lightning/src/chain/keysinterface.rs:228]: Did you say this policy enforcement has been implemented, namely "don't sign a revoked transaction" ? I think you need to verify the new `revoked_commitment` against commitment transaction number. I guess you can access it by unveiling it from `LocalCommitmentTransaction` 's unsigned_tx. This can be addressed later.

Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:228]: Oops, I misspoke above. I thought about implementing this, but didn't get to it.

File: lightning/src/util/enforcing_trait_impls.rs (hunk 2 above)
Comment (devrandom) [reply lightning/src/util/enforcing_trait_impls.rs]: So the implementation is incomplete, will fix that in a bit.

Comment (devrandom) [reply lightning/src/util/enforcing_trait_impls.rs]: I tried to do this, but I'm seeing some persistence issues.  I think using persistent fields in `EnforcingChannelKeys` doesn't work because there are multiple copies of the class around (likely via read/write), so the `revoked_commitment` field isn't staying updated as expected.

For now, I moved the third commit and the enforcement WIP commit to a new PR in our repo: https://github.com/lightning-signer/rust-lightning/pull/1

File: lightning/src/chain/keysinterface.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/chain/keysinterface.rs]: I'm not quite sure I follow why "revoke" is in the name. Isn't that dependent on the context in which the method is called?

I'd also rather not use the `get_` prefix for accessors as that isn't the convention.

**Comment by jkczyz** [comment]
Thanks for updating the PR description with the motivation. Could you also include this in the commit message as per the [contributing guidelines](https://github.com/rust-bitcoin/rust-lightning/blob/master/CONTRIBUTING.md#contribution-workflow)? This would be useful for those looking through the commit log.

Generally, I see the PR description used for the broader motivation for a set of commits, though here there is only one commit.

File: lightning/src/chain/keysinterface.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: Actually, after this PR getting the secret is *only* used for RevokeAndACK.  Let me know if I'm missing something.

File: lightning/src/chain/keysinterface.rs [hunk 4]
```diff
@@ -195,9 +195,16 @@ impl Readable for SpendableOutputDescriptor {
...
 // ChannelMonitors instead of expecting to clone the one out of the Channel into the monitors.
 pub trait ChannelKeys : Send+Clone {
-	/// Gets the commitment seed for a specific commitment number
-	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards
-	fn commitment_secret(&self, idx: u64) -> [u8; 32];
+	/// Gets the per-commitment point for a specific commitment number
+	///
+	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards.
+	fn get_per_commitment_point<T: secp256k1::Signing + secp256k1::Verification>(&self, idx: u64, secp_ctx: &Secp256k1<T>) -> PublicKey;
+	/// Gets the commitment seed for a specific commitment number, thereby revoking the commitment
```
Comment (valentinewallace) [inline lightning/src/chain/keysinterface.rs:202]: Is this line a little misleading right now? My understanding is that, right now, this function just returns the commitment secret for a specific commitment number, it doesn't actually do any revocation in itself.

File: lightning/src/chain/keysinterface.rs [hunk 5]
```diff
@@ -195,9 +195,16 @@ impl Readable for SpendableOutputDescriptor {
...
+	///
+	/// After this is called, it is an error to try to sign the relevant local commitment transaction.
```
Comment (valentinewallace) [inline lightning/src/chain/keysinterface.rs:204]: Is it actually an error right now, or are you saying that it's a bad idea? Could potentially be clarified

File: lightning/src/chain/keysinterface.rs (hunk 3 above)
Comment (valentinewallace) [reply lightning/src/chain/keysinterface.rs:227]: Maybe since it's not implemented yet, rephrase the line to be a TODO?

File: lightning/src/chain/keysinterface.rs (hunk 1 above)
Comment (jkczyz) [reply lightning/src/chain/keysinterface.rs]: Understood. But "revoke" describes how the caller (i.e., `Channel`) is using this function. The caller shouldn't need clarification (by including "revoke" in the name) on how it intends to use the result of a function that it calls.

Separately, I see the value of documenting this behavior for the implementor of the trait. In that respect, higher-level documentation on the trait itself would be most valuable rather than spreading across function-level documentation. But this is more of a critique of how the overall documentation of this trait is currently organized than the specific change in this PR.

Regarding my earlier comment on the `get_` prefix, I see your point that it's not a pure function. Though, I think using "get" doesn't imply that it's not either given that prefix is typically used for accessor methods in some languages. Given the in-memory implementation delegates to `build_commitment_secret`, perhaps "build" is more appropriate prefix.

File: lightning/src/ln/chan_utils.rs [hunk 6]
```diff
@@ -497,8 +497,8 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, to_self_del
 
 #[derive(Clone)]
 /// We use this to track local commitment transactions and put off signing them until we are ready
-/// to broadcast. Eventually this will require a signer which is possibly external, but for now we
-/// just pass in the SecretKeys required.
+/// to broadcast. This class can be used inside a signer implementation to generate a signature
+/// given the relevant secret key.
```
Comment (valentinewallace) [inline lightning/src/ln/chan_utils.rs:501]: Seems to be getting rid of some info here -- rather than losing it, maybe add a sentence at the end: "Eventually, a signer (whether external or internal) will be required, but for now the SecretKeys are passed in directly when need be."

File: lightning/src/chain/keysinterface.rs (hunk 5 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:204]: This is describing the contract for this interface method - i.e. that the caller must not try to sign this commitment.  It is up to the specific `ChannelKeys` (signer) implementation to actually enforce this, and `EnforcingChannelKeys` will do so soon.

File: lightning/src/chain/keysinterface.rs (hunk 4 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:202]: Since this is the interface, it just documents the desired contract.  The contract is that it's an error to try to sign a revoked transaction or revoke a signed transaction.  It is true to `EnforcingChannelKeys` doesn't *yet* implement this, so I agree that ECK should have a TODO reminding users of the class that it is incomplete.  I'll add the TODO.

File: lightning/src/ln/chan_utils.rs (hunk 6 above)
Comment (devrandom) [reply lightning/src/ln/chan_utils.rs:501]: `LocalCommitmentTransaction.get_local_sig` is now only called inside `ChannelKeys` implementations, so I believe the new wording is accurate.  I think the original wording is a holdover from the time that the secret key was passed out of the control of `ChannelKeys` implementations, but now the encapsulation of secret keys is complete due to the recent changes in that area by @ariard.

File: lightning/src/chain/keysinterface.rs (hunk 3 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:227]: Agreed, but again, I think it should be attached to `EnforcingChannelKeys`

File: lightning/src/chain/keysinterface.rs (hunk 4 above)
Comment (valentinewallace) [reply lightning/src/chain/keysinterface.rs:202]: I see. I'd rephrase it to: "Gets the commitment seed for a specific commitment number. This function is also responsible for revoking the commitment." Reasoning being, the word "thereby" seems implies that just by getting the commitment seed, revocation will result.

File: lightning/src/chain/keysinterface.rs (hunk 5 above)
Comment (valentinewallace) [reply lightning/src/chain/keysinterface.rs:204]: Ah OK -- I think RL's other docs tend to phrase "requirements that implementers of a trait must satisfy" a bit differently. To be consistent with our other docs, I'd rephrase to "After this method is called, the implementor of this trait must ensure that it will refuse sign the relevant local commitment transaction."

File: lightning/src/chain/keysinterface.rs (hunk 1 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: @jkczyz, you also came up with `release_commitment_secret`.  I like that one quite a bit, since it has an action verb and seems less cumbersome.

Comment (jkczyz) [reply lightning/src/chain/keysinterface.rs]: Yes, after some offline discussion @devrandom convinced me that -- from the point of view of the signer -- the commitment has been revoked. After some pondering, I thought `release_commitment_secret` seemed suitable since it signals the secret is in the wild but the actual revocation happens at a higher level of abstraction.

Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: done

File: lightning/src/chain/keysinterface.rs (hunk 3 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs:232]: done

File: lightning/src/chain/keysinterface.rs (hunk 5 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: rephrased, please take another look

File: lightning/src/chain/keysinterface.rs (hunk 4 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: I rephrased somewhat differently, please take a look

**Comment by devrandom** [comment]
I rephrased a few things and added TODOs.  Please take another look.

File: lightning/src/chain/keysinterface.rs [hunk 7]
```diff
@@ -195,9 +195,21 @@ impl Readable for SpendableOutputDescriptor {
...
-	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards
-	fn commitment_secret(&self, idx: u64) -> [u8; 32];
+	/// Gets the per-commitment point for a specific commitment number
+	///
+	/// Note that the commitment number starts at (1 << 48) - 1 and counts backwards.
+	fn get_per_commitment_point<T: secp256k1::Signing + secp256k1::Verification>(&self, idx: u64, secp_ctx: &Secp256k1<T>) -> PublicKey;
+	/// Gets the commitment secret for a specific commitment number as part of the revocation process
+	///
+	/// An implementation must error here if the commitment was already signed.
+	/// After this method is called, an implementation must ensure that it will refuse sign the relevant
```
Comment (TheBlueMatt) [inline lightning/src/chain/keysinterface.rs:205]: I don't think it MUST - in theory we ensure that on the rust-lightning end, certainly it can, at least if you're in a separate memory space :).

File: lightning/src/chain/keysinterface.rs (hunk 3 above)
Comment (TheBlueMatt) [reply lightning/src/chain/keysinterface.rs:232]: I believe all of our other comments are phrased as "what rust-lightning guarantees it will do" ie, instead of "The commitment must not have been revoked" this could read "The commitment transaction contained will not have been revoked by release_commitment_secret, and external signers may wish to enforce this".

Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: reworded, please take a look

File: lightning/src/chain/keysinterface.rs (hunk 7 above)
Comment (devrandom) [reply lightning/src/chain/keysinterface.rs]: reworded

**Comment by devrandom** [comment]
Done
